qutip/mesolve.py

```diff
@@ -510,22 +510,25 @@ def get_curr_state_data(r):
 
         if opt.store_states or expt_callback:
             cdata = get_curr_state_data(r)
+            fdata = dense2D_to_fastcsr_fmode(cdata, size, size)
 
-        if opt.store_states:
-            if issuper(rho0):
-                fdata = dense2D_to_fastcsr_fmode(cdata, size, size)
-                output.states.append(Qobj(fdata, dims=dims))
+            # Try to guess if there is a fast path for rho_t
+            if issuper(rho0) or not rho0.isherm:
+                rho_t = Qobj(fdata, dims=dims)
             else:
-                fdata = dense2D_to_fastcsr_fmode(cdata, size, size)
-                output.states.append(Qobj(fdata, dims=dims, fast="mc-dm"))
+                rho_t = Qobj(fdata, dims=dims, fast="mc-dm")
+
+        if opt.store_states:
+            output.states.append(rho_t)
 
         if expt_callback:
             # use callback method
-            output.expect.append(e_ops(t, Qobj(cdata, dims=dims)))
+            output.expect.append(e_ops(t, rho_t))
```

Comment:
if store_states=false, and exp_callback = true, rho_t is not defined to use here?

Reply:
Thanks for reviewing the PR Neill!. I think that if `expt_callback = True` and `store_states=False` the function should still work since it creates `rho_t` when any of those are True. However, I have just realized that I did not include any test for this case. I will include it as soon as possible.